### ISSUE TITLE
Istioctl Upgrade Fails on Non-Default Namespace

### ISSUE THREAD

**Issue by anton-yurchenko** [istio/istio#22198]
### Bug description:

istioctl v1.5.0 ignores custom namespaces in `config.yaml` and/or provided via argument `--istioNamespace=customSpace`.  

Executing `istioctl upgrade -f config.yaml` ends up like this:
```
proto: tag has too few fields: "-"
2020-03-15T16:38:32.890710Z	info	Error: failed to read the current Istio version, error: failed to retrieve Istio control plane version, error: istio pod not found in namespace istio-system
Error: failed to read the current Istio version, error: failed to retrieve Istio control plane version, error: istio pod not found in namespace istio-system
```

During 1 hour of work I have upgraded 1.4.4->1.4.5->1.4.6 with `istioctl experimental upgrade -f config.yaml` but executing the same command on 1.5.0 ended up with previously attached error log.
Providing `--istioNamespace=customSpace` to the command changed nothing and **istioctl** was still trying to work against a default namespace.

*P.S: pay attention that listing **Istio** version was successful.*
*P.S: I was not able to find that issue, sorry if I missed it.*

### Expected behavior:
Executing `istioctl upgrade -f config.yaml` should behave exactly like `istioctl experimental upgrade -f config.yaml` on previous versions.

### Steps to reproduce the bug:
Prepare a v1.4.x set up with some custom configurations and upgrade up to v1.5.0 including.

### Versions:
- Istio Version (collected by `istioctl --istioNamespace=customSpace version --remote` :exclamation:):
```
client version: 1.5.0
control plane version: 1.4.6
data plane version: 1.4.6 (1 proxies), 1.4.4 (14 proxies)
```

- Kubernetes Version:
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.2", GitCommit:"c97fe5036ef3df2967d086711e6c0c405941e14b", GitTreeState:"clean", BuildDate:"2019-10-15T23:42:50Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.10-eks-bac369", GitCommit:"bac3690554985327ae4d13e42169e8b1c2f37226", GitTreeState:"clean", BuildDate:"2020-02-26T01:12:54Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
```

- Helm Version (**Istio was NOT installed through Helm, but through istioctl**):
```
Client: &version.Version{SemVer:"v2.16.0", GitCommit:"e13bc94621d4ef666270cfbe734aaabf342a49bb", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.16.0", GitCommit:"e13bc94621d4ef666270cfbe734aaabf342a49bb", GitTreeState:"clean"}
```

### How was Istio installed?
`istioctl` with configuration YAML.

### Environment where a bug was observed (cloud vendor, OS, etc)
Kubernetes is on AWS EKS v1.15 and a developer workstation macOS Catalina Version 10.15.3.

### Sample from my config.yaml:
```yaml
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
spec:
  defaultNamespace: customSpace
  profile: default
  security:
    components:
      citadel:
        k8s:
          replicaCount: 2
  gateways:
    components:
      ingressGateway:
        enabled: true
  values:
    global:
      proxy:
        excludeIPRanges: 169.254.169.254/32
    prometheus:
      enabled: false
    gateways:
      istio-ingressgateway:
        loadBalancerSourceRanges:
          - 10.10.0.0/16
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-type: nlb
```

**Comment by so-jelly** [comment]
just to chime in with a very easy to reproduce example
```
istioctl manifest generate \
  --istioNamespace foo \
  --namespace bar |
    grep namespace: |
    sed 's/ *//' |
    sort |
    uniq
```

> namespace:
> namespace: istio-system

**Comment by esnible** [comment]
This has been fixed for 1.6.  Is there sufficient demand to get this into 1.5.x?

The fix for 1.6, https://github.com/istio/istio/issues/21778 , requires the user to use `--set namespace=<ns>`

There is a proposal to have `istioctl apply` record the configuration in the cluster so that it will be remembered by `istioctl upgrade` and `istioctl verify-install`.  Please vote for that feature if you want it.  https://github.com/istio/istio/issues/20000

**Comment by anton-yurchenko** [comment]
Thanks @esnible for a quick reply!

There is no specific need for an upgrade and we can definitely skip 1.5.x.

- Is that allowed to skip versions during an upgrade?
- When is 1.6 is expected to be released?

**Comment by anton-yurchenko** [comment]
As Istio 1.6 was recently released :tada: , I have tried to upgrade from 1.4.9 to 1.6.0 using: `istioctl upgrade --set namespace=customSpace -f config.yaml` and got the following error: 
`Error: upgrade version check failed: 1.4.9 -> 1.6.0. Error: upgrade is currently not supported: 1.4.9 -> 1.6.0` 

According to the [Upgrade Notes](https://istio.io/news/releases/1.6.x/announcing-1.6/upgrade-notes/), I must upgrade to 1.5 first, which in place does not accept a non-standard namespace.
> Currently, Istio doesn’t support skip-level upgrades. If you are using 
Istio 1.4, you must upgrade to Istio 1.5 first, and then upgrade to 
Istio 1.6.

I have tried to upgrading after setting `--enable-validation=false` on **istio-galley** (`kubectl edit deployment -n customSpace istio-galley`), without success.

@esnible, can you suggest any workaround for this maybe?

**Comment by esnible** [comment]
You may try using `istioctl upgrade --force`.  Using `--force` works for some settings, I don't know about this one.  (Galley validation is a different thing.)

You can also try using `istioctl manifest apply` to install, and then manually deleting, using `kubectl delete`), any 1.4 Deployments in _istio-system_ that were not cleaned up.  Also any ValidatingWebhookConfigurations.  I have not tried that for 1.4->1.6, if you go this route let us know how it works.

**Comment by esnible** [comment]
I should mention it is probably better to do the 1.4->1.5 and then 1.5->1.6.

Custom settings are not preserved during most upgrades.  Istio's point of view is that deploying Istio should be something that you do from a repeatable script or IstioOperator.  You don't want to be in a position where you only have a script for installing 1.4 with custom settings and then upgrading.

**Comment by anton-yurchenko** [comment]
I was using custom namespace because of this [issue](https://github.com/kubernetes/kubernetes/issues/69264) which is no more the case, so I ended up deleting 1.4.9 installation and installing 1.6.0 into an `istio-system` namespace.
Following your recommendation, I have moved all the configuration to `IstioOperator` except for the `Gateway` creation.

Thank you @esnible for your help and detailed explanations!
